File: BOJ/yein/9_week/게으른 백곰.py
N, K = map(int, input().split())

ice = {}
max_dist = 0
for _ in range(N):
    g, x = map(int, input().split())
    ice[x] = g
    max_dist = max(max_dist, x)

arr = []
for j in range(max_dist+1):
    if(j in ice): arr.append(ice[j])
    else: arr.append(0)

curr_sum = sum(arr[:min(2 * K + 1, max_dist + 1)])
curr_max = curr_sum

for k in range(1, max_dist + 1 - K):
    left = arr[k - 1] if k - 1 >= 0 else 0
    right = arr[k + 2 * K] if k + 2 * K <= max_dist else 0
    curr_sum = curr_sum - left + right
    curr_max = max(curr_max, curr_sum)

print(curr_max)

File: BOJ/jaehong/8_week/게으른 백곰.js
const fs = require("fs");
const input = fs
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")
  .map((line) => line.split(" ").map(Number));

const [N, K] = input.shift();
const coord = input;

function solution(K, coord) {
  let maxPosition = 0;
  const arr = Array(1000001).fill(0);

  for (const [ice, position] of coord) {
    arr[position] = ice;
    maxPosition = Math.max(maxPosition, position);
  }

  let windowSum = 0;
  for (let i = 0; i <= K && i <= maxPosition; i++) {
    windowSum += arr[i];
  }

  let maxIce = windowSum;
  for (let j = 0; j <= maxPosition; j++) {
    const left = j - K - 1;
    const right = j + K + 1;
    if (left >= 0) windowSum -= arr[left];
    if (right <= maxPosition) windowSum += arr[right];
    maxIce = Math.max(maxIce, windowSum);
  }

  return maxIce;
}

console.log(solution(K, coord));
